crates/walrus-sdk/src/client.rs

```diff
@@ -441,6 +510,172 @@ impl<T: ReadClient> Client<T> {
             result = future => result,
         }
     }
+
+    /// Retrieves slivers with retry logic, only requesting missing slivers in subsequent attempts.
+    ///
+    /// This function will keep retrying until all requested slivers are received or the maximum
+    /// number of retries is reached or the timeout is reached.
+    async fn retrieve_slivers_with_retry<E: EncodingAxis>(
+        &self,
+        metadata: &VerifiedBlobMetadataWithId,
+        sliver_indices: &[SliverIndex],
+        certified_epoch: Epoch,
+        max_attempts: Option<usize>,
+        timeout_duration: Option<Duration>,
+    ) -> Result<Vec<SliverData<E>>, ClientError>
+    where
+        SliverData<E>: TryFrom<walrus_core::Sliver>,
+    {
+        let mut sliver_selector = SliverSelector::<E>::new(
+            sliver_indices.to_vec(),
+            metadata.n_shards(),
+            metadata.blob_id(),
+        );
+        let mut attempts = 0;
+        let retry_delay_ms = 100;
+        let mut all_slivers: HashMap<SliverIndex, SliverData<E>> = HashMap::new();
+        let start_time = Instant::now();
+        let mut last_error: Option<ClientError> = None;
+
+        while !sliver_selector.is_empty() {
+            // Check if we've exceeded the timeout or the max retries.
+            if let Some(timeout_duration) = timeout_duration {
+                if start_time.elapsed() > timeout_duration {
+                    tracing::debug!(
+                        "Timeout reached after {:?} while retrieving slivers",
+                        timeout_duration
+                    );
+                    break;
+                }
+            }
+            if let Some(max_attempts) = max_attempts {
+                if attempts > max_attempts {
+                    tracing::debug!("Max attempts ({}) reached", max_attempts);
+                    break;
+                }
+            }
+
+            tracing::debug!("Retrieving slivers: {:?}", sliver_selector);
+            match self
+                .retrieve_slivers(metadata, &sliver_selector, certified_epoch)
+                .await
+            {
+                Ok(new_slivers) => {
+                    // Track which indices we've successfully retrieved.
+                    for sliver in new_slivers {
+                        sliver_selector.complete_sliver(&sliver.index);
+                        all_slivers.insert(sliver.index, sliver);
+                    }
+                }
+                Err(e) => {
+                    // TODO(heliu): if the error is not retriable, return the error immediately.
+                    tracing::warn!("Error retrieving slivers: {:?}", e);
+                    last_error = Some(e);
+                }
+            }
+
+            attempts += 1;
+            if !sliver_selector.is_empty() {
+                tokio::time::sleep(Duration::from_millis(retry_delay_ms)).await;
+            }
+        }
+
+        if all_slivers.len() != sliver_indices.len() {
```

Comment:
Important: This could also happen if `sliver_indices` contained some duplicates. We should document what happens in that case.